### PR TITLE
8318484: Initial version of cdsConfig.hpp

### PR DESCRIPTION
**Pull request by iklam**
This is the first step for [JDK-8318483 - Move CDS configuration management into cdsConfig.hpp](https://bugs.openjdk.org/browse/JDK-8318483)

- Remove `Arguments::is_dumping_archive()` and `Arguments assert_is_dumping_archive()`
- Add the following new APIs

```
class CDSConfig {
  static bool is_dumping_archive();
  static bool is_dumping_static_archive();
  static bool is_dumping_dynamic_archive();
  static bool is_dumping_heap();
};
```

- Convert some use of `DumpSharedSpaces` and `DynamicDumpSharedSpaces` to these new APIs

(More APIs will be added in future sub tasks of [JDK-8318483](https://bugs.openjdk.org/browse/JDK-8318483))

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318484](https://bugs.openjdk.org/browse/JDK-8318484): Initial version of cdsConfig.hpp (**Sub-task** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16257/head:pull/16257` \
`$ git checkout pull/16257`

Update a local copy of the PR: \
`$ git checkout pull/16257` \
`$ git pull https://git.openjdk.org/jdk.git pull/16257/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16257`

View PR using the GUI difftool: \
`$ git pr show -t 16257`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16257.diff">https://git.openjdk.org/jdk/pull/16257.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16257#issuecomment-1770156262)

### CODE REVIEW

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 1]
```diff
@@ -774,10 +775,12 @@ void MetaspaceShared::preload_and_dump_impl(TRAPS) {
   log_info(cds)("Rewriting and linking classes: done");
 
 #if INCLUDE_CDS_JAVA_HEAP
-  StringTable::allocate_shared_strings_array(CHECK);
-  ArchiveHeapWriter::init();
-  if (use_full_module_graph()) {
-    HeapShared::reset_archived_object_states(CHECK);
+  if (CDSConfig::is_dumping_heap()) {
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/metaspaceShared.cpp:778]: This seems a new condition. Why is it needed now?

**Review by dholmes-ora — APPROVED** [review]
Initial refactoring looks good. One query below.

Thanks

File: src/hotspot/share/cds/metaspaceShared.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/metaspaceShared.cpp:778]: This was a bug uncovered during refactoring. `StringTable::allocate_shared_strings_array()` used to assert `DumpSharedSpaces`. However, this function is useful only in a more limited scope (`CDSConfig::is_dumping_heap()` which is a subset of `DumpSharedSpaces`). So after changing the assert in  `StringTable::allocate_shared_strings_array()`, I have to change the condition where this function is called.

File: src/hotspot/share/cds/cdsConfig.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+class CDSConfig : public AllStatic {
+public:
+  // Basic CDS features
+  static bool      is_dumping_archive()                      NOT_CDS_RETURN_(false);
+  static bool      is_dumping_static_archive()               NOT_CDS_RETURN_(false);
+  static bool      is_dumping_dynamic_archive()              NOT_CDS_RETURN_(false);
+
+  // CDS archived heap
+  static bool      is_dumping_heap()                         NOT_CDS_JAVA_HEAP_RETURN_(false);
```
Comment (calvinccheung) [inline src/hotspot/share/cds/cdsConfig.hpp:39]: Too much blank spaces between the function declarations and NOT_CDS_* macros.
The function declarations could also be shifted more to the left.

**Review by calvinccheung — APPROVED** [review]
One nit in cdsConfig.hpp.

File: src/hotspot/share/cds/cdsConfig.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsConfig.hpp:39]: The spaces are for functions that will be added in the next PR that have longer names:

```
  // Basic CDS features
  static bool      is_dumping_archive()                      NOT_CDS_RETURN_(false);
  static bool      is_dumping_static_archive()               NOT_CDS_RETURN_(false);
  static bool      is_dumping_dynamic_archive()              NOT_CDS_RETURN_(false);

  // CDS archived heap
  static bool      is_dumping_heap()                         NOT_CDS_JAVA_HEAP_RETURN_(false);
  static bool      is_loading_heap()                         NOT_CDS_JAVA_HEAP_RETURN_(false);
  static void disable_dumping_full_module_graph(const char* reason = nullptr) NOT_CDS_JAVA_HEAP_RETURN;
  static bool      is_dumping_full_module_graph()            NOT_CDS_JAVA_HEAP_RETURN_(false);
  static void disable_loading_full_module_graph(const char* reason = nullptr) NOT_CDS_JAVA_HEAP_RETURN;
  static bool      is_loading_full_module_graph()            NOT_CDS_JAVA_HEAP_RETURN_(false);


```

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

**Comment by iklam** [comment]
Thanks @dholmes-ora @calvinccheung @sspitsyn for the review.
/integrate

**Pull request closed without merging**